File: LeetCode_ProblemSet/Problem2873.cpp
#include <algorithm>
#include <vector>

class Solution {
   public:
    long long maximumTripletValue(std::vector<int>& nums) {
        int n = nums.size();
        std::vector<int> leftMax(n);
        std::vector<int> rightMax(n);
        for (int j = 1; j < n; j++) {
            leftMax[j] = std::max(leftMax[j - 1], nums[j - 1]);
            rightMax[n - 1 - j] = std::max(rightMax[n - j], nums[n - j]);
        }
        
        long long answer = 0;
        for (int j = 1; j < n - 1; j++) {
            answer = std::max(
                answer,
                (long long)(leftMax[j] - nums[j]) * rightMax[j]
            );
        }
        return answer;
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem2873.rs
impl Solution {
    pub fn maximum_triplet_value(nums: Vec<i32>) -> i64 {
        let n = nums.len();
        let mut left_max = vec![0; n];
        let mut right_max = vec![0; n];
        for j in 1..n {
            left_max[j] = std::cmp::max(left_max[j - 1], nums[j - 1]);
            right_max[n - 1 - j] = std::cmp::max(right_max[n - j], nums[n - j]);
        }
        let mut answer = 0;
        for j in 1..(n - 1) {
            answer = std::cmp::max(
                answer,
                (left_max[j] - nums[j]) as i64 * right_max[j] as i64
            );
        }
        return answer;
    }
}